### PR TITLE
Added preferred name to users API

### PR DESCRIPTION
**Pull request by carlosfelix2**
## Summary
Fetches the demographics and adds the preferred name to the profile. This also updates some tests that use the profile.

## Related issue(s)
https://app.zenhub.com/workspaces/mhv-on-vagov-landing-page-62619a987d74510018ecc546/issues/gh/department-of-veterans-affairs/va.gov-team/74398

## Testing done
- unit tests added and updated
- Manual testing using Demographics mock data (added via [this PR](https://github.com/department-of-veterans-affairs/vets-api/pull/15972) and [this PR](https://github.com/department-of-veterans-affairs/vets-api-mockdata/pull/463))
- See also [this PR](https://github.com/department-of-veterans-affairs/vets-api-mockdata/pull/464) for added mock data that can be used with mocked local authentication

Manual testing:
1. Checkout the `cfelix/74398-demographics-mock` branch for the `vets-api-mockdata` repo.
1. Checkout the branch for this PR for the `vets-api`
1. Start the `vets-api` locally
1. Start the `vets-website` locally
1. Access the [local mock login](http://localhost:3001/sign-in/mocked-auth)
1. Open the browser's developer tools and preserve the logs for the Network tab.
1. Use the mock login and login as an IDME user.
1. In the browser's developer tools network tab, find the call to the user's API (`user`) and verify that the preferred name is present. Note there is a default mock response for all users as well as a few specific responses in [this PR](https://github.com/department-of-veterans-affairs/vets-api-mockdata/pull/464).

## Screenshots
N/A

## What areas of the site does it impact?
- v0/user API

## Acceptance criteria

- [x]  I fixed|updated|added unit tests and integration tests for each feature (if applicable).
- [x]  No error nor warning in the console.
- [x]  Events are being sent to the appropriate logging solution
- [x]  Documentation has been updated (link to documentation)
- [x]  No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs
- [x]  Feature/bug has a monitor built into Datadog or Grafana (if applicable)
- [x]  If app impacted requires authentication, did you login to a local build and verify all authenticated routes work as expected
- [x]  I added a screenshot of the developed feature

## Requested Feedback
This makes a new request when fetching the profile via the v0/user API. We added this call to be performed only when the user API gets a request and kept it out of the user model since it is not used anywhere else. Is there a preference here for this call to be in the user model, and if so, wouldn't that create a performance issue since user is used everywhere?

### CODE REVIEW

File: spec/services/users/profile_spec.rb [hunk 1]
```diff
@@ -288,12 +298,13 @@
         end
 
         it 'populates the #errors array with the serialized error', :aggregate_failures do
-          error = subject.errors.first
-
-          expect(error[:external_service]).to eq 'MVI'
-          expect(error[:start_time]).to be_present
-          expect(error[:description]).to include 'Not authorized'
-          expect(error[:status]).to eq 401
+          VCR.use_cassette('va_profile/demographics/demographics') do
```
Comment (carlosfelix2) [inline spec/services/users/profile_spec.rb:301]: Note: This is a different cassette as the user ID is different.

File: spec/services/users/profile_spec.rb [hunk 2]
```diff
@@ -515,5 +540,20 @@
           .to eq({ auth_broker: SAML::URLService::BROKER_CODE, ssoe: true, transactionid: 'a' })
       end
     end
+
+    describe 'preferred name' do
```
Comment (carlosfelix2) [inline spec/services/users/profile_spec.rb:544]: Note: these are the new unit tests

File: spec/requests/user_request_spec.rb [hunk 3]
```diff
@@ -15,8 +15,10 @@
       create(:account, idme_uuid: mhv_user.uuid)
       sign_in_as(mhv_user)
       allow_any_instance_of(User).to receive(:edipi).and_return(edipi)
-      VCR.use_cassette('va_profile/veteran_status/va_profile_veteran_status_200', allow_playback_repeats: true) do
-        get v0_user_url, params: nil, headers: v0_user_request_headers
+      VCR.use_cassette('va_profile/demographics/demographics', allow_playback_repeats: true) do
+        VCR.use_cassette('va_profile/veteran_status/va_profile_veteran_status_200', allow_playback_repeats: true) do
+          get v0_user_url, params: nil, headers: v0_user_request_headers
+        end
       end
```
Comment (radavis) [inline spec/requests/user_request_spec.rb:22]: Instead of nesting calls to `VCR.use_cassette`, I would suggest the following:

```
    cassettes = [
      { name: 'profile/demographics' },
      { name: 'va_profile/veteran_status/va_profile_veteran_status_200' }
    ]
    VCR.use_cassettes cassettes do
      get v0_user_url, params: nil, headers: v0_user_request_headers
    end
```

An alternate way, utilizing the rspec `around` block could be:

```
 around do |test_case|
    cassettes = [
      { name: 'profile/demographics', allow_playback_repeats: true },
      { name: 'va_profile/veteran_status/va_profile_veteran_status_200', match_requests_on: %i[method body], allow_playback_repeats: true }
    ]
    VCR.use_cassettes cassettes do
      test_case.run
    end
  end
```

Note that you can add configuration for the cassette within the hash for that cassette. Typically you would only want to use `allow_playback_repeats: true` if the test case will make more than one request for the same resource.

File: spec/support/vcr_cassettes/profile/demographics.yml [hunk 4]
```diff
@@ -0,0 +1,56 @@
+---
+http_interactions:
+- request:
+    method: get
+    uri: https://int.vet360.va.gov/demographics/demographics/v1/2.16.840.1.113883.4.349/9d018700-b72c-444a-95b4-43e14a4509ea%5EPN%5E200VIDM%5EUSDVA
```
Comment (radavis) [inline spec/support/vcr_cassettes/profile/demographics.yml:5]: Could this file live in the same folder as `spec/support/vcr_cassettes/va_profile/demographics/demographics.yml`, since they both are recordings of the API interactions with `https://int.vet360.va.gov/demographics/demographics/v1`?

**Review by radavis — CHANGES_REQUESTED** [review]
Left a few suggestions. Let me know if any of these asks are more trouble than they are worth.

File: spec/requests/user_request_spec.rb [hunk 5]
```diff
@@ -202,13 +206,15 @@
...
 
-      expect(response.status).to eq 296
-      expect(error['external_service']).to eq 'MVI'
-      expect(error['description']).to be_present
-      expect(error['status']).to eq 401
+        expect(response.status).to eq 296
+        expect(error['external_service']).to eq 'MVI'
+        expect(error['description']).to be_present
+        expect(error['status']).to eq 401
+      end
```
[the remark this answers is not in the document]
Comment (carlosfelix2) [reply spec/requests/user_request_spec.rb:217]: Good one. Removed

File: spec/requests/user_request_spec.rb (hunk 3 above)
Comment (carlosfelix2) [reply spec/requests/user_request_spec.rb:22]: 1. Interesting that I could not find any documentation anywhere on how to do this. I tried it out, but could not get it working. I found info that says the preference is to use the blocks anyways and if not use the insert/eject method, but that method is discouraged as it can lead to bad states in tests.
2. As far as the `allow_playback_repeats`, I wanted to match what was on the other cassette. I was able to remove some of it, but some tests do make multiple requests because they have multiple tests in one test.

File: spec/support/vcr_cassettes/profile/demographics.yml (hunk 4 above)
Comment (carlosfelix2) [reply spec/support/vcr_cassettes/profile/demographics.yml:5]: It can live anywhere, but I thought the location was based on the test and not the content of the cassette. Can you confirm?

File: app/services/users/profile.rb [hunk 6]
```diff
@@ -63,12 +64,21 @@ def account
       nil
     end
 
+    def demographics
```
Comment (carlosfelix2) [inline app/services/users/profile.rb:67]: Note: Note the location of this call is here instead of user since we only need the demographics data when the user endpoint is called vs the user object is used in many other places.

**Review by jzucadi — APPROVED** [review]

File: app/services/users/profile.rb [hunk 7]
```diff
@@ -63,12 +64,21 @@ def account
...
+      nil
+    end
+
     def profile
       {
         email: user.email,
         first_name: user.first_name,
         middle_name: user.middle_name,
         last_name: user.last_name,
+        preferred_name: demographics&.preferred_name&.text,
```
Comment (dickdavis) [inline app/services/users/profile.rb:81]: Chaining safe navigation operators like this is generally not a good practice and can be a code smell. It can make things harder to read or understand where the nil values are coming from.

For instance, if this value were to be blank, does that mean...
* the call to the demographics API failed?
* the user doesn't have a preferred name?
* the value for preferred name doesn't respond to the `text` message being sent?

I looked at the underlying service and the associated response classes, and I'm not too certain there's much you can do here to address my concern.

All that being said, I think this particular value should be added to the `mpi_profile` return value instead. [This line](https://github.com/department-of-veterans-affairs/vets-api/blob/0253f4695a40f9640e1a3bc41a9d8d1b4c3e754f/app/services/users/profile.rb#L52) seems to extract all of the VA profile data from that method.

Comment (carlosfelix2) [reply app/services/users/profile.rb:81]: @dickdavis Thank you for the feedback.
1. I think I can wrap some of that logic into my new function and properly log a failure. I will try that out.
2. The problem I have with adding that value to the `mpi_profile` is that it has nothing to do with VA Profile even when the name of the attribute in the scaffold is called `va_profile`. All the values in that section come from MPI, and `mpi_profile` even has an error catcher for the MPI service. This is the reason I placed the preferred name in the `profile` section. I can still just add `preferred name` to `mpi_profile` as my new function will catch the error properly, but I just feel we are mixing sources there. Of course, I am saying all this with no real history on what was intended for each of those functions. Does it make sense to keep that separate, or I am reading too much into how the code should work?

File: spec/support/vcr_cassettes/profile/demographics.yml (hunk 4 above)
Comment (carlosfelix2) [reply spec/support/vcr_cassettes/profile/demographics.yml:5]: Location of VCR cassettes tends to be by the test that uses them. I think this location matches that pattern.

File: app/services/users/profile.rb (hunk 7 above)
Comment (dickdavis) [reply app/services/users/profile.rb]: Ha, I think we are probably both in the same boat with regard to knowing the original intention.

I think I have an idea. What if we tried...

* Add `demographics` to the scaffold [here](https://github.com/department-of-veterans-affairs/vets-api/blob/master/app/services/users/scaffold.rb#L11)
```ruby
class Scaffold < Struct.new(... :demographics, ...)
```

* Update the `fetch_and_serialize_profile` method [here](https://github.com/department-of-veterans-affairs/vets-api/blob/ea8f21d88045970d5f8830ec96bf5577efcddbcc/app/services/users/profile.rb#L48) to scaffold an entirely new spot for demographics. This should make it cleaner to add additional data pulled through calls using the demographics service in the future. Something like this:
```ruby
    def fetch_and_serialize_profile
      ...
      scaffold.demographics = demographics_data
      ...
    end
```

* Then, implement the `demographics_data` method (name it whatever you like really):
```ruby
    def demographics_data
      demographic_svc = VAProfile::Demographics::Service.new @user
      # We want to catch if preferred_name or text is missing, so no safe operator here.
      data = demographic_svc.get_demographics.demographics
      {
        preferred_name: data.preferred_name.text
      }
    rescue => e
      scaffold.errors << Users::ExceptionHandler.new(e, 'VAProfile').serialize_error
      nil
    end
```

* Add the `demographics` attribute to the `UserSerializer` [here](https://github.com/department-of-veterans-affairs/vets-api/blob/ea8f21d88045970d5f8830ec96bf5577efcddbcc/app/serializers/user_serializer.rb#L9) and delegate the method call to the object somewhere around [here](https://github.com/department-of-veterans-affairs/vets-api/blob/ea8f21d88045970d5f8830ec96bf5577efcddbcc/app/serializers/user_serializer.rb#L25)
```ruby
  attributes ... :demographics, ...

  ...

  delegate :demographics, to: :object
```

**Review by radavis — DISMISSED** [review]
Approving changes requested.

**Review by jzucadi — APPROVED** [review]
nice work!

**Comment by dickdavis** [comment]
I think we are super close to landing this one. The last thing to do is add coverage for the demographics attribute in the [UserSerializer spec](https://github.com/department-of-veterans-affairs/vets-api/blob/master/spec/serializers/user_serializer_spec.rb). I'll go ahead and pull this branch down to test the authentication flow.

**Review by dickdavis — CHANGES_REQUESTED** [review]
Ok, I ran into issues testing this out locally since betamocks wasn't set up for the demographics service yet. I opened up two PRs that need to go in before this one so we can test this and ensure auth doesn't break.

https://github.com/department-of-veterans-affairs/vets-api/pull/15972
https://github.com/department-of-veterans-affairs/vets-api-mockdata/pull/463

File: app/services/users/profile.rb [hunk 8]
```diff
@@ -63,6 +65,22 @@ def account
       nil
     end
 
+    # Get the preferred named from the user's demographics. See scaffold.errors for any errors.
+    #
+    # @return the user's preferred name or nil if the user has no preferred name or there was an
+    # error fetching the name.
+    #
+    def demographics_data
+      demographic_svc = VAProfile::Demographics::Service.new @user
```
Comment (dickdavis) [inline app/services/users/profile.rb:74]: Minor detail I overlooked previously, but we can change this to reference `user` instead of `@user` since we are using the `attr_reader` to define the getter method. Not a super big deal, but since the rest of the class uses `user` we might as well stay consistent.

File: config/betamocks/services_config.yml [hunk 9]
```diff
@@ -866,4 +866,8 @@
   :endpoints:
     - :method: :get
       :path: "/demographics/demographics/v1/*/*"
-      :file_path: "vet360/demographics/default"
+      :file_path: "vet360/demographics"
+      :cache_multiple_responses:
+        :uid_location: url
+        # Supports UIDs with or without dashes
+        :uid_locator: '\/demographics\/demographics\/v1\/[0-9.]+\/([0-9A-Fa-f-]+)'
```
Comment (carlosfelix2) [inline config/betamocks/services_config.yml:873]: Note: This change allows us to use different mock data based on the user ID.

**Review by dickdavis — APPROVED** [review]
LGTM, thanks for incorporating my suggestions!

File: app/services/users/profile.rb [hunk 10]
```diff
@@ -63,6 +65,22 @@ def account
...
+      demographics = demographic_svc.get_demographics.demographics
+      {
+        preferred_name: demographics.preferred_name.text
```
Comment (LindseySaari) [inline app/services/users/profile.rb:77]: Do you want to use safe method chaining here or are you just letting the rescue take care of any issues with obtaining demographic info?

Comment (carlosfelix2) [reply app/services/users/profile.rb:77]: We wanted the rescue to take care of that, so we can catch any API issues like text is not supported, etc.

File: spec/requests/user_request_spec.rb [hunk 11]
```diff
@@ -10,12 +10,20 @@
...
+      ]
+    end
+
     before do
       allow_any_instance_of(MHVAccountTypeService).to receive(:mhv_account_type).and_return('Premium')
       create(:account, idme_uuid: mhv_user.uuid)
       sign_in_as(mhv_user)
       allow_any_instance_of(User).to receive(:edipi).and_return(edipi)
-      VCR.use_cassette('va_profile/veteran_status/va_profile_veteran_status_200', allow_playback_repeats: true) do
+      VCR.use_cassettes(cassettes_good_vet_status) do
```
Comment (LindseySaari) [inline spec/requests/user_request_spec.rb:26]: Why the change to the cassette name here?

Comment (LindseySaari) [reply spec/requests/user_request_spec.rb:26]: I would also like to keep the directory nesting and naming scheme if possible

Comment (carlosfelix2) [reply spec/requests/user_request_spec.rb:26]: That cassette is still being used, but we are not adding the demographics cassette too. So, two cassettes are being used.

File: spec/requests/user_request_spec.rb [hunk 12]
```diff
@@ -177,7 +185,7 @@
       it 'returns meta.errors information', :aggregate_failures do
         error = body.dig('meta', 'errors').first
 
-        expect(error['external_service']).to eq 'VAProfile'
+        expect(error['external_service']).to be_present
```
Comment (LindseySaari) [inline spec/requests/user_request_spec.rb:188]: Why does this no longer equal VAProfile?

File: spec/services/users/profile_spec.rb [hunk 13]
```diff
@@ -8,15 +8,35 @@
   let!(:in_progress_form_user_uuid) { create(:in_progress_form, user_uuid: user.uuid) }
   let!(:in_progress_form_user_account) { create(:in_progress_form, user_account: user.user_account) }
 
+  let(:cassettes_good_vet_status) do
```
Comment (LindseySaari) [inline spec/services/users/profile_spec.rb:11]: Same comment here

File: spec/requests/user_request_spec.rb (hunk 11 above)
Comment (carlosfelix2) [reply spec/requests/user_request_spec.rb:26]: Directory nesting is also kept. See like 15 and 16 for the cassettes being used here.

Comment (LindseySaari) [reply spec/requests/user_request_spec.rb:26]: Oh! I'm sorry! I overlooked that. I see the definition above

File: spec/requests/user_request_spec.rb (hunk 12 above)
Comment (carlosfelix2) [reply spec/requests/user_request_spec.rb:188]: The test originally expected the first error (and only error) to be from the VA Profile veteran status call. After the changes on the PR, the test will generate two errors, one for the demographics and one for the veteran status. I changed it just to verify data is there since the test is only testing for the existence of error metadata in the response.

File: spec/services/users/profile_spec.rb (hunk 13 above)
Comment (carlosfelix2) [reply spec/services/users/profile_spec.rb:11]: This is not a cassette name, it is a variable name. The cassettes are in line 13 and 14.

**Review by LindseySaari — APPROVED** [review]
